File: flutter/omise/example/lib/viewmodel/omise_viewmodel.dart
import 'dart:convert';

import 'package:example/model/omise_model.dart';
// ignore: depend_on_referenced_packages
import 'package:http/http.dart' as http;


class OmiseViewModel {

  String redirectUrl = "";
  Map<String, dynamic> dicOmise = {};
  bool isProgress = false;

  Future<void> createCharge(int intAmount
                            , String strName
                            , String strCardNumber
                            , String expMonth
                            , String expYear
                            , String securityCode
                            , String strReturnUrl
                            , String strPaymentId
                              ) async{
    http.Response objResponse = await OmiseModel.createCharge(intAmount, strName, strCardNumber, expMonth, expYear, securityCode, strReturnUrl, strPaymentId);
    dicOmise = jsonDecode(objResponse.body);
                            
    return;
  }

  Future<void> paypayCharge(int intAmount, String strReturnUrl, String strPaymentId ) async{

    http.Response objResponse = await OmiseModel.chargePayPay(intAmount, strReturnUrl, strPaymentId);
    dicOmise = jsonDecode(objResponse.body);
    redirectUrl = (dicOmise["redirectUri"] ?? "").toString();
    return;
  }

  Future<void> paypayChargeAnother(int intAmount, String strReturnUrl, String strPaymentId ) async{

    http.Response objResponse = await OmiseModel.chargePayPayAnother(intAmount, strReturnUrl, strPaymentId);
    dicOmise = jsonDecode(objResponse.body);
    redirectUrl = (dicOmise["redirectUri"] ?? "").toString();
    return;
  }

  Future<void> fetchCharge(String strPaymentId) async{

    http.Response objResponse = await OmiseModel.fetchCharge(strPaymentId);
    dicOmise = jsonDecode(objResponse.body);
    return;
  }

}

File: swiftui/OpnPaymentExample/OpnPaymentExample/OmiseViewModel.swift
//
//  OmiseViewModel.swift
//  OpnPaymentExample
//
//  Created by Shunsuke Hiratsuka on 2023/10/05.
//

import Foundation
import OmiseSDK

final class OmiseViewModel: ObservableObject, Equatable {
    
    static func == (lhs: OmiseViewModel, rhs: OmiseViewModel) -> Bool {
        if lhs.isProgress != rhs.isProgress || lhs.strRedirectUrl != rhs.strRedirectUrl{
            return true
        }else{
            return false
        }
        
    }
    
    
    private let model = OmiseModel()
    @Published var isProgress: Bool = false
    @Published var strRedirectUrl: String = ""
    @Published var chargeId: String = ""
    @Published var dicOmise: [String: Any]?
    @Published var actionError: Error?
    
    func postCreditCardCharge(param intAmount: Int, strName: String, strCardNumber: String, expMonth: String, expYear: String, secCode: String, strReturlUrl: String, strPaymentId: String) async throws {
        
        model.setupClient()
        try await model.chargeCreditCard(param: intAmount, strName: strName, strCardNumber: strCardNumber, expMonth: expMonth, expYear: expYear, securityCode: secCode, strReturnUrl: strReturlUrl, strPaymentId: strPaymentId, completion: completionCharge)
        
    }
    
    func postPayPayCharge(param intAmount: Int, strReturnUrl: String, strPaymentId: String) async throws {
        
        try await model.chargePayPay(param: intAmount, strReturnUrl: strReturnUrl, strPaymentId: strPaymentId, completion: completionPayPay)
        
    }
    
    func postPayPayAnother(param intAmount: Int, strReturnUrl: String, strPaymentId: String) async throws {
        
        model.setupClient()
        try await model.chargePayPayAnother(param: intAmount, strReturnUrl: strReturnUrl, strPaymentId: strPaymentId, completion: completionPayPay)
        
    }
    
    
    func fetchChargeInfo(param strPaymentId: String) async throws {
        
        try await model.fetchPayPay(param: strPaymentId, completion: completionPayPay)
    }
    
    
    func completionPayPay(data objData: [String : Any]?, error: Error?) {
        
        actionError = error
        
        self.dicOmise = objData
        
        let objRedirectUri = self.dicOmise?["redirectUri"] ?? ""
        
        self.strRedirectUrl = String(describing: objRedirectUri)
        
    }
    
    func completionCharge(data objData: [String : Any]?, error: Error?) {
        
        actionError = error
        
        self.dicOmise = objData
        
        let objChargeId = self.dicOmise?["id"] ?? ""
        
        self.chargeId = String(describing: objChargeId)
        
    }
    
    func completionFetch(data objData: [String : Any]?, error: Error?) {
        
        actionError = error
        
        self.dicOmise = objData
        
    }
    
}
